### PR TITLE
Snowflake test incremental without state

### PR DESCRIPTION
**Pull request by Ismail-Naziz**
## What
test in incremental without initial state

## How
all 4 possible types of states are managed

## Review guide
changes on source code to correct the behavior in edge cases
test_table.py in unittest integration tests

## Can this PR be safely reverted and rolled back?
 YES 💚

### CODE REVIEW

File: airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py [hunk 1]
```diff
@@ -184,10 +185,10 @@ def setUp(self):
             "parent_stream": f"{_SCHEMA}.{_TABLE}",
             "where_clause": self.where_clause
         }
-        self.record = a_snowflake_response("response_get_table")
+        self.get_record_builder_instance = lambda: a_snowflake_response("response_get_table")
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py:188]: Nice catch on seeing that we were modifying the same instance of the builder without creating a new one

nit: renaming this `self._a_record` would reduce noise when reading tests

File: airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py [hunk 2]
```diff
@@ -292,3 +292,219 @@ def _read(cls,
...
+    @HttpMocker()
+    def test_incremental_without_initial_state_with_date_cursor(self, uuid_mock, mock_auth, http_mocker: HttpMocker) -> None:
+        _given_get_timezone(http_mocker)
+        _given_read_schema(http_mocker)
+        _given_table_catalog(http_mocker)
+        _given_table_with_primary_keys(http_mocker)
+
+        cursor_field = "TEST_COLUMN_20"
+        id_index = 12
+        expected_cursor_value = "1970-01-04"
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py:366]: Can we explain why `1970-01-04` is the expected value? Checking the records that are created [here](https://github.com/artefactory/airbyte/pull/37/files#diff-84ce19b13999dd0a0a5e7a0de9c7a4869deb78b1e04abb954161704c4b028b95R386-R388) does not explain clearly why we are passing from "3" to this date

**Review by maxi297 — APPROVED** [review]
I wasn't able to run the tests because of the freezegun import. It does not seem to be used. Can we clean this import and the others that are not used as well? 

Conditional approval on fixing:
* nits (naming and the link between record value and expection)
* the imports

File: airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py (hunk 2 above)
Comment (Ismail-Naziz) [reply airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py:366]: in snowflake api, for date type specifically, snowflake sends the number of days (positive or negative) from the date 1970-1-1 meaning that 3 is 1970-1-1 + 3 days gives 1970-1-4

File: airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py (hunk 1 above)
Comment (Ismail-Naziz) [reply airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py]: Done

File: airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py (hunk 2 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py:355]: Cool! I'm fine with either adding a comment to explain that or use the code to document that. Something like:

```
most_recent_cursor = datetime(1970, 1, 4)
expected_cursor_value = most_recent_cursor.strftime("%Y-%m-%d")
```
and `(most_recent_cursor - datetime(1970, 1, 1)).days` to set the cursor value at line `377`

Comment (Ismail-Naziz) [reply airbyte-integrations/connectors/source-snowflake/unit_tests/integration/test_table.py:355]: Concerning documentation of this behavior, I prefer to document it at the place where the transformation happens to avoid duplication of explanation of behavior. I will put a comment to refer to the place where it is documented in the test. The transformation happens in this function source_snowflake.schema_builder.format_field. The next PR will concern a unittest about this with behavior documented for every case.

**Review by maxi297 — APPROVED** [review]

File: airbyte-integrations/connectors/source-snowflake/unit_tests/integration/request_builder.py [hunk 3]
```diff
@@ -72,6 +74,15 @@ def with_where_statement(self, where_statement: str):
         self._where_statement = where_statement
         return self
 
+    def with_state(self, state):
```
[the remark this answers is not in the document]
Comment (Ismail-Naziz) [reply airbyte-integrations/connectors/source-snowflake/unit_tests/integration/request_builder.py:77]: corrected in this PR https://github.com/artefactory/airbyte/pull/45 because the code is refactored. No more concepts of state and cursor in request builder.

File: airbyte-integrations/connectors/source-snowflake/source_snowflake/schema_builder.py [hunk 4]
```diff
@@ -150,12 +150,15 @@ def format_field(field_value, field_type, local_time_zone_offset_hours=None):
...
 
-    if field_type.upper() in date_and_time_snowflake_type_airbyte_type.keys() and field_value:
+    if field_type.upper() in date_and_time_snowflake_type_airbyte_type.keys() and field_value is not None:
 
         if isinstance(field_value, datetime):  # Already formatted date
             return field_value
 
         try:
+            if not isinstance(field_value, str) and len(field_value) == 0:
+                raise ValueError
```
[the remark this answers is not in the document]
Comment (Ismail-Naziz) [reply airbyte-integrations/connectors/source-snowflake/source_snowflake/schema_builder.py:160]: Yes I will correct that in PR 38 that is dedicated to schema builder

**Comment by Ismail-Naziz** [comment]
All the comments are adjusted eiither in this PR or next PRs, can you approve this one ?

**Comment by Ismail-Naziz** [comment]
Already merged
